bpf/bpf_host.c

```diff
@@ -1156,15 +1161,28 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
 
 		identity = resolve_srcid_ipv6(ctx, ip6, identity, &ipcache_srcid, from_host);
 		ctx_store_meta(ctx, CB_SRC_LABEL, identity);
-		if (from_host) {
+
 # if defined(ENABLE_HOST_FIREWALL) && !defined(ENABLE_MASQUERADE_IPV6)
+		if (from_host) {
 			/* If we don't rely on BPF-based masquerading, we need
 			 * to pass the srcid from ipcache to host firewall. See
 			 * comment in ipv6_host_policy_egress() for details.
 			 */
 			ctx_store_meta(ctx, CB_IPCACHE_SRC_LABEL, ipcache_srcid);
+		}
 # endif /* defined(ENABLE_HOST_FIREWALL) && !defined(ENABLE_MASQUERADE_IPV6) */
+
+#ifdef ENABLE_WIREGUARD
+		if (!from_host) {
+			next_proto = ip6->nexthdr;
+			l4_off = ETH_HLEN + ipv6_hdrlen(ctx, &next_proto);
+			if (ctx_is_wireguard(ctx, l4_off, next_proto, ipcache_srcid))
+				trace.reason = TRACE_REASON_ENCRYPTED;
```

Comment:
@learnitall with https://github.com/cilium/cilium/pull/29616 in mind for context, do we need to add `trace.monitor=0` here? If yes, then shouldn't this line be caught by `encrypt_agg.cocci`? As far as I understand it's not triggering the cocci patch because there is no explicit `trace.monitor=${something-non-zero}` following this line.

Reply:
I had the [same concern](https://github.com/cilium/cilium/issues/34659#:~:text=The%20encrypt_agg.cocci%20has%20been%20introduced%20to%20make%20sure%20that%20no%20aggregation%20is%20performed%20when%20tracing%20encrypted%20packet%20(IPSec).%20However%2C%20when%20applying%20the%20patch%20hence%20tracing%20Ingress%20Wireguard%20as%20encrypted%2C%20the%20cil_from_netdev%20program%20uses%20the%20TRACE_PAYLOAD_LEN%20macro.). Differently than IPSec, these traces would be generated anyway independently by us marking with TRACE_REASON_ENCRYPTED or not. For this reason, here I didn't want to loose the TRACE_PAYLOAD_LEN value.
I also agree that what you describe is the reason why  `encrypt_agg.cocci` is not detecting this. Let me know in case anything else is needed.

Reply:
> shouldn't this line be caught by `encrypt_agg.cocci`?

It _should_ but this usage pattern of `trace_ctx` wasn't added. We can look into adding another case which requires that `trace.monitor = 0` follows `trace.reason = TRACE_REASON_ENCRYPTED`.
 
> Differently than IPSec, these traces would be generated anyway independently by us marking with TRACE_REASON_ENCRYPTED or not.

Can you expand on what you mean? I understand that the traces would be generated anyway regardless of if the packet was WG or not, but I don't understand what you mean by 'differently than IPSec'.

Overall, the reason why we've paired setting `monitor=0` with `TRACE_REASON_ENCRYPTED` is we've assumed that anytime a `TRACE_REASON_ENCRYPTED` event is sent, the packet the event was generated for lacks the connection tracking information to perform event aggregation, and as a result, these events produce a lot of noise. My understand of the specifics of the aggregation techniques and how we handle WG traffic isn't developed enough to provide a 'yes' or 'no' here, so I'd recommend we experiment:

1. Can we compare the observed events for WG traffic before and after the PR? If this PR dramatically increases the number of events, then we should set `monitor=0`. For example, here's a before/after comparison of the effect of https://github.com/cilium/cilium/pull/29616 for IPSec using `hubble observe --follow`:

Before:

```
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) -> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) <> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) <> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.069: default/test2:53538 (ID:41889) -> kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-endpoint FORWARDED (UDP)
Feb 22 22:21:59.070: default/test2:53538 (ID:41889) <- kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-stack FORWARDED (UDP)
Feb 22 22:21:59.071: kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) <> default/test2:53538 (ID:41889) to-stack FORWARDED (UDP)
Feb 22 22:21:59.071: kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) <> default/test2:53538 (ID:41889) to-stack FORWARDED (UDP)
Feb 22 22:21:59.071: default/test2:53538 (ID:41889) <- kube-system/kube-dns-6f9b8847ff-9d6kz:53 (ID:58873) to-endpoint FORWARDED (UDP)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: SYN)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: SYN)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: SYN)
Feb 22 22:21:59.072: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:21:59.073: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.073: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.074: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.076: default/nginx:80 (ID:33458) <> default/test2:49594 (ID:41889) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) <- default/nginx:80 (ID:33458) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:21:59.076: default/test2:49594 (ID:41889) -> default/nginx:80 (ID:33458) to-stack FORWARDED (TCP Flags: ACK)
```

After:

```
Feb 22 22:26:47.377: default/test:47138 (ID:40856) -> kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-stack FORWARDED (UDP)
Feb 22 22:26:47.377: default/test:47138 (ID:40856) -> kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-endpoint FORWARDED (UDP)
Feb 22 22:26:47.378: default/test:47138 (ID:40856) <- kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-stack FORWARDED (UDP)
Feb 22 22:26:47.378: default/test:47138 (ID:40856) <- kube-system/kube-dns-6f9b8847ff-hqbpb:53 (ID:28365) to-endpoint FORWARDED (UDP)
Feb 22 22:26:47.379: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: SYN)
Feb 22 22:26:47.379: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: SYN)
Feb 22 22:26:47.379: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: SYN, ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.380: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.381: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.381: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
Feb 22 22:26:47.383: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.383: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) <- default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK, FIN)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-stack FORWARDED (TCP Flags: ACK)
Feb 22 22:26:47.392: default/test:40960 (ID:40856) -> default/nginx:80 (ID:8979) to-endpoint FORWARDED (TCP Flags: ACK)
```

2. Can we try setting `monitor=0` here and observe the impact on the generated events? If there's information missing or the impact isn't noticeable, then we can pass on it.

@smagnani96 would you have time to try these tests out? I can help out as well if you'd like.

Reply:
@learnitall I talked with Simone and he helped me understand that changing the trace reason for wg in this PR shouldn't affect monitor aggregation.

The two changes in `do_netdev` (this thread and [L1223](https://github.com/cilium/cilium/pull/35183/files/785a3c9f94dab5457371db8c308b839c40153f2f#r1795537643)) are in the ingress path, and the observation point will consistently be `TRACE_FROM_NETWORK` in the wg case. `TRACE_FROM_*` events will not be emitted unless monitor aggregation is set to "none" _regardless of the `monitor` value_, see [here](https://github.com/cilium/cilium/blob/253a05bef38306624eb6675c16a4e90cb3e5c729/bpf/lib/trace.h#L165-L177).

[`cil_to_netdev`](https://github.com/cilium/cilium/pull/35183/files/785a3c9f94dab5457371db8c308b839c40153f2f#r1795543959) is in the egress path, but there we go through[ `handle_to_netdev_ipv{6,4}`](https://github.com/cilium/cilium/blob/253a05bef38306624eb6675c16a4e90cb3e5c729/bpf/bpf_host.c#L1443-L1454) and do the ct lookup.

Thus, I marked my comments as resolved, but feel free to respond if you think we've missed something :pray:

Reply:
Ok sweet! That makes a ton of sense to me. Thank you!

Reply:
Thanks to you both :smiley: 
As mentioned by @kaworu, this patch shouldn't lead to logging more events than we would normally expect:

1. ingress: `TRACE_FROM_NETWORK` is not logged by default when `monitorAggregation > none`
2. egress: `trace.monitor` is modified only by a successful `ct_lookup`, which shouldn't happen when dealing with the encrypted packet, thus not logging the event when `monitorAggregation > none`.

For instance, with the default `monitorAggregation=medium` I see only the following packets:

```bash
#request
-> stack flow 0x0 , identity 19646->4691 state new ifindex 0 orig-ip 0.0.0.0: 10.244.1.178 -> 10.244.2.58 EchoRequest
-> 0: 10.244.1.178 -> 10.244.2.58 EchoRequest
#response
-> endpoint 388 flow 0x0 , identity 4691->19646 state reply ifindex lxc2e2524a80f9c orig-ip 10.244.2.58: 10.244.2.58 -> 10.244.1.178 EchoReply
```

Reply:
>egress: trace.monitor is modified only by a successful ct_lookup, which shouldn't happen when dealing with the encrypted packet, thus not logging the event when monitorAggregation > none.

This is related to this code section, right?
https://github.com/cilium/cilium/blob/a2cf84ec76daa6c12950f7f0efa7948d7d904f30/bpf/bpf_host.c#L1527

There's two aspects to consider
1. when the HostFW is enabled, the preceding Egress-policy check would have already performed a CT lookup (since encrypted WG traffic looks like host-originating traffic), and touched the `trace` struct. Arguably we should be smarter here, and as https://github.com/cilium/cilium/issues/32335 suggests not apply HostFW to our own WG traffic (or are we already doing that? Should be straight-forward to repro).
2. when BPF masq is enabled, we apply CT for [host-originating traffic like WG](https://github.com/cilium/cilium/issues/34089) (to avoid port collisions with SNATed pod-originating connections). Here handle_nat_fwd() would touch the `trace` struct.

Reply:
Hey @julianwiedmann, thanks for the feedback!
Your comment made me realize that right now we might lose the encryption bit in that codepath, such as in case of `handle_nat_fwd()`. I'm trying to tackle those cases in #35354, any further comment is appreciated :smiley: